lib/github.com/diku-dk/sorts/radix_sort.fut

```diff
@@ -109,3 +109,182 @@ def radix_sort_float_by_key [n] 't 'k
   by_key_wrapper radix_sort_float key num_bits get_bit xs
+
+local def exscan op ne xs =
+  let s =
+    scan op ne xs
+    |> rotate (-1)
+  let s[0] = ne
+  in s
+
+local def get_bin 't
+                  (get_bit: i32 -> t -> i32)
+                  (digit_n: i32)
+                  (x: t): i64  =
+  i64.i32 <| get_bit (digit_n+1) x * 2 + get_bit digit_n x
+
+local def radix_sort_step_i16 [n] 't (xs: [n]t)
+                              (get_bit: i32 -> t -> i32)
+                              (digit_n: i32): ([n]t, (i16, i16, i16, i16)) =
+  let num x = i16.i32 (get_bit (digit_n+1) x * 2 + get_bit digit_n x)
+  let pairwise op (a1,b1,c1,d1) (a2,b2,c2,d2) =
+    (a1 `op` a2, b1 `op` b2, c1 `op` c2, d1 `op` d2)
+  let bins = xs |> map num
+  let flags = bins |> map (\x ->
+      ( i16.bool (x==0)
+      , i16.bool (x==1)
+      , i16.bool (x==2)
+      , i16.bool (x==3) ) )
+  let offsets = scan (pairwise (+)) (0,0,0,0) flags
+  let (na,nb,nc,nd) = last offsets
+  let f bin (a,b,c,d) = i64.i16 ((-1)
+      + a * (i16.bool (bin == 0)) + na * (i16.bool (bin > 0))
+      + b * (i16.bool (bin == 1)) + nb * (i16.bool (bin > 1))
+      + c * (i16.bool (bin == 2)) + nc * (i16.bool (bin > 2))
+      + d * (i16.bool (bin == 3))) 
+  let is = map2 f bins offsets
+  in (scatter (copy xs) is xs, (na, nb, nc, nd))
+
+local def chunked_radix_sort_step [n] [m] [r] 't
+                          (get_bit: i32 -> t -> i32)
+                          (digit_n: i32)
+                          (xs: *[n*m+r]t) =
+  let hist_size = 4
+  let sort_step  arr =
+    let (ys, (a, b, c, d)) =
+      radix_sort_step_i16 arr get_bit digit_n
+    let hist' = sized hist_size [i64.i16 a
+                                ,i64.i16 b
+                                ,i64.i16 c
+                                ,i64.i16 d]
+    in (ys, hist')
+  let (chunks, rest) = split xs
+  let (sorted_rest, hist_rest) =
+    if r == 0
+    then ([], sized hist_size [0, 0, 0, 0])
+    else sort_step rest
+  let (sorted_chunks, hist_chunks) =
+    unflatten chunks
+    |> map sort_step
+    |> unzip
+  let histograms = hist_chunks ++ [hist_rest]
+  let ys = flatten sorted_chunks ++ sorted_rest
+  let flat_trans_hist =
+    histograms
+    |> transpose
+    |> flatten
+  let flat_hist =
+    histograms
+    |> flatten
+    |> sized (hist_size * (n + 1))
+  let (flat_trans_hist_scan, flat_hist_scan) =
+    zip flat_trans_hist flat_hist
+    |> exscan (\(a, b) (x, y) -> (a + x, b + y)) (0, 0)
+    |> unzip
+  let hist_scan =
+    flat_hist_scan
+    |> sized ((n + 1) * hist_size)
+    |> unflatten
+  let trans_hist_scan =
+    flat_trans_hist_scan
+    |> unflatten
+  let (is, elems) =
+    map (
+      \i ->
+        let elem = ys[i]
+        let bin = get_bin get_bit digit_n elem
+        let new_offset = trans_hist_scan[bin][i / m]
+        let old_offset = hist_scan[i / m][bin]
+        let idx = (i - old_offset) + new_offset
+        in (idx, elem)
+    ) (iota (n * m + r))
```

Comment:
Use `tabulate` instead of mapping over `iota`.

Reply:
This should be done now.